src/coreclr/vm/interoplibinterface_comwrappers.cpp

```diff
@@ -349,15 +356,55 @@ namespace
 
             // Insert objects into enumerable.
             // The ExternalObjectContexts in the hashmap are only
-            // removed and associated objects collected during a GC. Since
-            // this code is running in Cooperative mode they will never
-            // be null.
+            // detached from the cache during a GC. Since this code is
+            // running in Cooperative mode they will not be null during
+            // detection above, but could be marked for collection
+            // after allocating the above array.
+            SIZE_T objCount = 0;
             for (SIZE_T i = 0; i < localList.Size(); i++)
             {
                 ExternalObjectContext* inst = localList[i];
-                gc.arrRef->SetAt(i, inst->GetObjectRef());
+                if (!inst->IsSet(ExternalObjectContext::Flags_Detached))
+                {
+                    gc.arrRef->SetAt(objCount, inst->GetObjectRef());
+                    objCount++;
+                }
+                else
+                {
+                    // The context has been detach for this cache.
+                    // Since the state of this object is now under
+                    // the control of the Finalizer/GC, we null it out. This
+                    // allows us to ignore it when we iterate over the
+                    // collection later. The null here is needed because
+                    // it indicates the array size is going to change
+                    // and a new array allocated. That allocation is an
+                    // opportunity for the GC to remove it from this hashmap
+                    // and free the external object context.
```

Comment:
```suggestion
                    // and a new array will be allocated. That allocation is
                    // an opportunity for the GC to remove the object from
                    // this hashmap and free the external object context.
```

Reply:
Actually this seems a bit off:

> The null here is needed because it indicates the array size is going to change

That's based on the count, not this nulling out, no?

If I'm reading the comment, right, maybe something like:

In this scenario, the array size is going to change and a new array allocated will be allocated. Because we null out the object here, that allocation of the new array is an opportunity for the GC to remove the object from this hashmap and free the external object context.

Reply:
Sure. Will update with a new comment. GitHub is being difficult.